### DISCUSSION TITLE
Dual Recordings

### DISCUSSION THREAD

**Discussion by jesusmartin-cortecero** [petersenpeter/CellExplorer#72] [Q&A] [answered]
Hi Peter

I usually record two different regions in total 128 channles (Cortex H3 probe = 64 channels, and Thalamus H5/E1 probe =64 channels).
After recording i separately run kilosort for each recorded region (as far as i am aware Kilosort doesnt allow you to introduce two different channel maps). In the next step i would like to use Cell Explorer to calculate the Connectivity matrix and connectivity Graph between these two regions.  I would like to get the cell_metrics of each region independently and later analysis the connectivity between these two structures in different epochs. 

As far as i know Cell explorer reads the "rez.mat" for kilosort data, as i have two rez.mat files (one per structure) i dont know how to provide this information in the Session-gui. 

is it possible to calculate the conectivity graph between two different regions recorded with two different probes with CellExplorer?

Best 

Jesus

**Comment by petersenpeter** [comment]
Hi Jesus

Are the two regions recorded in the same raw binary file? If so I would create one channelmap for both probes.

In Kilosort you can define the two probe layouts as a single channelmap. Check our Kilosortwrapper for examples: https://github.com/petersenpeter/KilosortWrapper
This script generates the channelmap: https://github.com/petersenpeter/KilosortWrapper/blob/master/createChannelMapFile_KSW.m

In general, CellExplorer operates with one spikes struct. So if you have two rez files, what you could do is import them separately into two spikes structs, and then merge the spikes structs into one, before running CellExplorer or the detection of the monosynaptic connections. You would need to write a script for merging the spikes structs yourself.

**Comment by jesusmartin-cortecero** [comment]
Hi Peter thanks for your fast reply

I didnt explore these scripts yet.

I tried first to use a bin file with 128 channels and I then created a channel map with 2 probes H5 Cortex and E1 thalamus but it seems Cell Explorer is considering both as the same Probe.  
i made sure that the 64 channels of E1 are designated as kcoords =1 and H5 as kcoords =2 

https://drive.google.com/file/d/1cGCltNQNDPtD2P2UlOZ5QOh7fSniMjfM/view?usp=sharing %channel map

looking like this 

![image](https://user-images.githubusercontent.com/53047456/194562506-afe2fa96-9741-4aa1-8437-c1df8ff16658.png).
when i run kilosort i can see that the probes look good and the spike templates position are restricted to the specified kcoords. 
![image](https://user-images.githubusercontent.com/53047456/194564897-eb0b88c3-63d6-4d67-9471-a60137a22159.png)

however when i run CellExplorer i get the following trilatered position
![image](https://user-images.githubusercontent.com/53047456/194561824-599d8bb3-2758-4329-bebc-c861da2d751a.png). 

Is there an additional parameter that i need to set up in the session-gui before running the Cell_metrics gui?

Thank you very much in advance

**Comment by petersenpeter** [comment]
Your channel map looks fine, except for the large offset between the two probes. Perhaps you need to define the bad channels (e.g. noisy or floating channels)? They can skew the estimation of cell's spatial locations... they are stored in: session.channelTags.Bad.channels, but you can also add them via the GUI.

Otherwise, could you try to delete all the files generated by CellExplorer (session, spikes, cell metrics, and so on), and rerun the ProcessCellMetrics?

**Comment by petersenpeter** [comment]
Hi Jesus. Could you share the session with me? I can much better troubleshoot this way.

**Reply by jesusmartin-cortecero** [reply]
Hi peter, 
When i opened the file with NeuroScope2 i realized that there were channels duplicated, Then I deleted the files and i am doing everything from scratch, including the bin file, just to make sure everything is correct. Asap i will share with you the new session file.

**Reply by jesusmartin-cortecero** [reply]
Hi peter, the bin file was wrong and i had duplicated 64 channels, finally i could fix it, and i could run  phy on it to check everything was looking good. 
![image](https://user-images.githubusercontent.com/53047456/194669994-1a0549e1-badd-4910-b4d1-cb39fd4cbed0.png)
also in NeuroScope2
![image](https://user-images.githubusercontent.com/53047456/194670809-f2b001df-a17e-41ee-8eab-df66ef5bfd54.png)


however, now i cannot load the spikes with Cell Explorer 

spikes=loadSpikes('session',session, 'labelsToRead',{'good','mua'})
loadSpikes: Loading Phy data
Dot indexing is not supported for variables of this type.

Error in loadSpikes (line 313)
            disp(['Importing ' num2str(numel(spikes.times)),'/',
            num2str(length(dataArray{1})),' clusters from phy'])

is the first time it happens this error, and dont know how to proceed. 

https://drive.google.com/file/d/1wQpC5YOTLZR7D9DSECBOhkqtZEh4HYKk/view?usp=sharing  %this is the rez file 
https://drive.google.com/file/d/1XBr1DmL5gWn-EqbfAFDMcy_kT8nLB7Ph/view?usp=sharing     %session.mat

Thank you very much in advance

**Comment by jesusmartin-cortecero** [comment]
Finally seems to be working. i just run again kilosort and the problem seems to be fixed. 
although it would be nice to know what can trigger the error in loading spikes to prevent it in the future. 

this is how it looks like without manual curation i guess the 3-4 clusters in between the 2 probes might be due to electric artifacts. 
![image](https://user-images.githubusercontent.com/53047456/194676168-6bc90a68-c8ed-4213-a8e0-8a82a37d4211.png)

**Comment by petersenpeter** [comment]
Could you share your binary file with me as well?

**Reply by petersenpeter** [reply]
When extracting the spikes, you should be able to see if artifacts are contaminating your data. But first share your raw data (binary file) and I should be able to figure it out.

**Comment by jesusmartin-cortecero** [comment]
Hi Peter thank you very much for all your help. 

I saw few noisy clusters while CellExplorer was loading the spikes. I did not curate the clusters with kilosort to do the test.  
 
https://1drv.ms/u/s!AtWyEKzSiztLmjdovp6k-XBhQyvw?e=Q2pV54
Here is the Bin file of the dual recording test i was using to get familiar with cell explorer.  
https://drive.google.com/file/d/1NltYnCg2_adzR5bYKFdJlXEiDCnEaSkH/view?usp=sharing  %cellInfo
https://drive.google.com/file/d/1QyZ7wYaTF4SZsZyaO0u1EvyCK9Xwk-IT/view?usp=sharing  %session
https://drive.google.com/file/d/1e0Fu8ZdjDJIBOxMZ8arywpynp4gVsDSy/view?usp=sharing %channel info
https://drive.google.com/file/d/1bSRW8Z8YyAyFTu6FPIngzDTuRV9gjzIZ/view?usp=sharing %monosyn
https://drive.google.com/file/d/1D2tfqmG9JIq7611IRzBvOgk6EqQf2ENu/view?usp=sharing %cellMetrics
https://drive.google.com/file/d/1ImmEqzF4zqCiCqu6jvD7-05bQLXZDvuv/view?usp=sharing %rez.mat

Where i could find more information about how cell explorer calculates the trilatered position and how it calculates de connectivity graph?

**Comment by petersenpeter** [comment, marked as answer]
Hi Jesus
Sorry for the late reply. I had a chance to look at your data today. Are you doing stimulation - it seems like you have some large artifacts?

I created an updated session struct, where I added channel 15 as a Bad channel (It has huge artifacts), and created 4 electrode groups instead of one. Check it out. Now the cells are correctly projected on your shanks/probes. There are two cells that are shown in the middle of the probes, but I don't think they are real neurons and appear more to be related to noise artifacts. 

[temp_wh.session.mat.zip](https://github.com/petersenpeter/CellExplorer/files/9800409/temp_wh.session.mat.zip)

The trilaterated position is calculated in this script: https://github.com/petersenpeter/CellExplorer/blob/40e7c11d531f1673bf569d34d5e7236c95ec3f6f/calc_CellMetrics/trilat.m

It uses the spike amplitudes from the channels with the largest amplitudes (default up to 16 channels) and estimates the position using trilateration (check the code).

The connectivity graph is estimated from the monosynaptic connections. It does not know the strength of the connections, purely based on the binary connectivity matrix. I am using Matlab's built-in plot: https://se.mathworks.com/help/matlab/ref/graph.conncomp.html

**Reply by jesusmartin-cortecero** [reply]
Hi Peter, 
Thank you very much. 
when i run  cell_metrics script, i see 
 "Loading previous detected MonoSynaptic connections
  Determining transmission probabilities"

also in the cell_metrics structure (cell_metrics.putativeConnections) i can see that cellexplorer calculates excitatory/inhibitory transmission 
probability, i wonder what those values account for. 
![image](https://user-images.githubusercontent.com/53047456/197970008-fab21b6e-ab54-45e1-aac6-5d2227755a0d.png)
![image](https://user-images.githubusercontent.com/53047456/197970083-2e2da288-1c11-486f-adc9-883d2420b749.png)



Thanks again

**Reply by petersenpeter** [reply]
Hi Jesus

CellExplorer always loads existing metrics (basename.cellmetrics.cellinfo.mat), so it does not need to calculate all metrics again. So be aware of this when troubleshooting.

Regarding the synaptic strength, you can inspect the code here:
https://github.com/petersenpeter/CellExplorer/blob/487d5f1efce7f2448ef72de34ef0b82d363ceb3b/calc_CellMetrics/ce_GetTransProb.m

**Reply by rajatsaxena** [reply]
For the synaptic strength calculation, my understanding is it is the sum of values in the integration window (0.8-0.48ms). For excitatory connections, a larger value will indicate stronger synaptic strength. But for inhibitory connections, does a smaller value indicate more inhibition since the sum is smaller? 
Thank you so much.

**Reply by petersenpeter** [reply]
Hmm, I believe you are correct, this calculation is only correct for excitatory connections. Let me look into this and get back to you

**Reply by petersenpeter** [reply]
I gave this some thought. These metrics are transmission probabilities (not transmission strength as previously stated). The probability that a spike in the presynaptic cell will result in a spike in the postsynaptic cell. This makes sense for excitatory connections but is not suitable for inhibitory connections...

**Reply by rajatsaxena** [reply]
That makes sense. Any suggestions/ research article on then, what would be a good way to quantify strength of I-E connections? Thank you

**Reply by McKenzieNeuro** [reply]
Hello,

Here are some references:
https://www.nature.com/articles/nn.2134
https://www.nature.com/articles/s42003-022-03450-5
https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005738

However, I don't think these can appropriately be inferred from spike timing especially in CA1. Here are some words of caution.
1) when we stimulate  PV+ interneurons, we can clearly see inhibition in CA1 PYR firing rate that for some reason is not well captured in the CCG between spontaneous PV spikes and PYR spikes. This could be partially because the PYR rate is low, but there is something missing - a modeler should tackle this problem. 
2) in the cortex I have seen nice dips in the CCGs between INT and PYR like in Fujisawa 2008, maybe this is real inhibition, but it has never been validated as I did with Dan English https://pubmed.ncbi.nlm.nih.gov/29024669/
3) The interneuon network is super synchronous, we very often see 0-lag millisecond timescale synchrony between pairs of cells (it is the rule). Therefore, it is very difficult (impossible?) to ascribe pairwise relations since INT spikes tend to occur with other INT spikes.

In short - I strongly recommend not making a causal anatomical inference from this correlations.

Sam

**Reply by petersenpeter** [reply]
Thanks, Sam!
Indeed for CA1, you should be cautious about making claims about inhibitory connections. I have seen very strong dips in the CCGs (inhibitory connections) in other regions.

**Reply by rajatsaxena** [reply]
Thank you both for your comments! Definitely agree with both of you. I'm currently recording from Visual and motor cortex and have definitely seen very strong dips in CCGs. I'm really intrigued by the #1 raised by Sam, will think about it more. Thanks once again for the prompt reply.
